### ISSUE TITLE
feat(datasets): scoring dataset runs, e.g. precision, recall, f-value

### ISSUE THREAD

**Issue by marcklingen** [langfuse/langfuse#2512]
### Discussed in https://github.com/orgs/langfuse/discussions/2511

<div type='discussions-op-text'>

<sup>Originally posted by **yusuke-intern** July  2, 2024</sup>
### Describe the feature or potential improvement

In LangSmith, for example, there is a feature to get precision, recall, f-value in experiment feature on a dataset.
https://docs.smith.langchain.com/how_to_guides/evaluation/evaluate_existing_experiment
Can we have similar feature in Langfuse?

### Additional information

_No response_</div>

**Comment by yusuke-intern** [comment]
@marcklingen 
Thank you for your prompt response.
Is it currently impossible to calculate a single score for multiple traces and display it on the UI?, such as precision, recall, f-value?

**Comment by marcklingen** [comment]
Yes, this is currently not supported. You could add this to the metadata of a run (which is displayed on the dataset runs table) but it is not natively supported by the scoring capabilities

**Comment by marcklingen** [comment]
Thanks again for raising this!

**Comment by yusuke-intern** [comment]
> Yes, this is currently not supported. You could add this to the metadata of a run (which is displayed on the dataset runs table) but it is not natively supported by the scoring capabilities

Sounds great idea! thank you!!

**Comment by dillon-tribe** [comment]
@marcklingen It seems like this would be done by adding DatasetRunItem with the desired run_metadata, but I think it's best to add these stats after all run items are already created. It seems like I can hackily make this work by using item.link(..., trace_id="fake_trace_id_for_updating_run_metadata"), but could you instead support updating dataset run metadata directly, independently of adding a new dataset run item?

**Comment by justinas-kazanavicius** [comment]
This one is a crucial feature for my team - some metrics are just impossible to calculate per prediction so they have to be done at a dataset level. Metadata approach is not ideal as this way you cannot see the metrics in the dataset run page as separate columns, and having it all in one place rather than split across 2 would be much better. 

**Bonus ideas:** it would also be useful if we could also add a comment to each of the dataset metrics like you can have one at a dataset item level, for example expanding a comment for precision/recall (averages) could show precision/recall at a class level. Also, at the moment the comments get truncated with a small number of characters at the dataset level and they don't support markdown or anything to make tables with, so maybe the max size could be expanded and markdown support added.

**Comment by dan-robinson-ai** [comment]
+1

**Comment by marliessophie** [comment]
Hi @niclasvaneyk-teaminternet, thanks for reaching out. As of v3 the python SDK supports posting dataset run level scores. Please upgrade your instance to use this feature. Do not hesitate to reach out should you encounter more issues. 

Keeping this issue open as we do not natively support calculating precision, recall, f-value within Langfuse.

**Comment by lehoanglong95** [comment]
Hi @marliessophie, could you please give me example how to update dataset run level scores?

**Comment by marliessophie** [comment]
This is currently not possible. Would this be useful? We could consider supporting in UI update of dataset run level scores. If you have some context on your use case this would be helpful.

**Comment by panagiotisrit** [comment]
Hi, Would it be a possibility to add a small example on the python SDK level?
Currently its unclear to me how I would be able to link an aggregated measure (like f1 score) to a dataset run. I am assuming that @lehoanglong95 was referring to the same thing. 

Thank you!

**Comment by anujsinha3** [comment]
Hi, 
Upvoting on the last comment by @panagiotisrit, can you please add an SDK example of how to update run-level score.
Not able to find the option in the code, or in the documentation to implement it.

**Comment by thomasgtaylor** [comment]
Hello! Also upvoting the last comment to receive an update from the team or a code snippet on how to do this.

**Comment by thomasgtaylor** [comment]
@anujsinha3 @panagiotisrit @lehoanglong95 

I figured it out 😄. If you want to add a score to the dataset run-level as an aggregate, here's how in Python:

```python
langfuse = get_client()
run_name = "some-run-name"
dataset_name = "some-dataset-name"
dataset = langfuse.get_dataset(name="some-dataset-name")
scores = []
for item in dataset.items:
    with item.run(
        run_name=run_name
    ) as root_span:
        thread_id = str(uuid4())
        result = app.invoke(
            {
                "messages": [HumanMessage(item.input)]
            },
            {
                "configurable": {
                    "thread_id": thread_id,
                },
                "metadata": {
                    "langfuse_session_id": thread_id,
                },
                "callbacks": [callback_handler],
            }
        )
        root_span.update(input=item.input, output=result)
        score = int(result["output"] == item.expected_output)
        scores.append(score)
        root_span.score(
            name="some-score-name",
            value=score,
            data_type="BOOLEAN",
        )

# after all the run items have been scored
# get the dataset run by name to get the id
run = langfuse.api.datasets.get_run(
   dataset_name=data.name,
   run_name=run_name
)
langfuse.create_score(
    dataset_run_id=run.id,
    name="some_aggregate",
    value=sum(scores) / len(scores),
    data_type="NUMERIC"
)
langfuse.flush()
```

**Comment by NoamGit** [comment]
> [@anujsinha3](https://github.com/anujsinha3) [@panagiotisrit](https://github.com/panagiotisrit) [@lehoanglong95](https://github.com/lehoanglong95)
> 
> I figured it out 😄. If you want to add a score to the dataset run-level as an aggregate, here's how in Python:
> 
> langfuse = get_client()
> run_name = "some-run-name"
> dataset_name = "some-dataset-name"
> dataset = langfuse.get_dataset(name="some-dataset-name")
> scores = []
> for item in dataset.items:
>     with item.run(
>         run_name=run_name
>     ) as root_span:
>         thread_id = str(uuid4())
>         result = app.invoke(
>             {
>                 "messages": [HumanMessage(item.input)]
>             },
>             {
>                 "configurable": {
>                     "thread_id": thread_id,
>                 },
>                 "metadata": {
>                     "langfuse_session_id": thread_id,
>                 },
>                 "callbacks": [callback_handler],
>             }
>         )
>         root_span.update(input=item.input, output=result)
>         score = int(result["output"] == item.expected_output)
>         scores.append(score)
>         root_span.score(
>             name="some-score-name",
>             value=score,
>             data_type="BOOLEAN",
>         )
> 
> # after all the run items have been scored
> # get the dataset run by name to get the id
> run = langfuse.api.datasets.get_run(
>    dataset_name=data.name,
>    run_name=run_name
> )
> langfuse.create_score(
>     dataset_run_id=run.id,
>     name="some_aggregate",
>     value=sum(scores) / len(scores),
>     data_type="NUMERIC"
> )
> langfuse.flush()

This works! thanks!

**Comment by flawnn** [comment]
Did anybody figure out a clean way to add run-level dataset scores? I have F1 scores, and adding one score for each category is kind of verbose and fills up the scores really quickly.

And any nice way to visualize metrics across different runs as well? 👀
